### PR TITLE
Add MuTransfer Capablity to RETRO model pretraining

### PR DESCRIPTION
**Pull request by yidong72**
# What does this PR do ?

It adds the muTransfer capablity to do RETRO model pretraining.
1. Included the mutransfer codes from MS github.
2. Add shape calculation script.
3. Add the mutransfer pretraining script.
4. Make sure it uses fused_softmax padding mode because of the customized masks
5. Include a CI test to make sure the model training performance matches. 

It also fix a bug that PTL grad sync is not called.

To use muTransfer for other models (GPT, T5, BERT etc), you can follow the same pattern.
1. register muadam or muadamw optimizer.
2. swap to use MuReadout to to convert model_d to logistics
3. Use the similar parameters init method in the mutransfer_pretrain file.

### CODE REVIEW

**Review by MaximumEntropy — COMMENTED** [review]
Just a minor comment

File: examples/nlp/language_modeling/conf/megatron_retro_mutransfer.yaml [hunk 1]
```diff
@@ -0,0 +1,293 @@
+name: test_retro
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_retro_mutransfer.yaml]: fixed

File: examples/nlp/language_modeling/conf/megatron_retro_mutransfer.yaml [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+  ffn_hidden_size: 384 # Transformer FFN hidden size. Usually 4 * hidden_size.
+  num_attention_heads: 12
+  init_method_std: 0.02 # Standard deviation of the zero mean normal distribution used for weight initialization.')
+  hidden_dropout: 0.1 # Dropout probability for hidden state transformer.
+  attention_dropout: 0.1 # drop out probability for attention matrix
+  kv_channels: null # Projection weights dimension in multi-head attention. Set to hidden_size // num_attention_heads if null
+  apply_query_key_layer_scaling: True # scale Q * K^T by 1 / layer-number.
+  layernorm_epsilon: 1e-5
+  gradient_as_bucket_view: True # Allocate gradients in a contiguous bucket to save memory (less fragmentation and buffer memory)
+  persist_layer_norm: False
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_retro_mutransfer.yaml]: no particular reason. It works with True.

**Comment by yidong72** [comment]
@MaximumEntropy , I imported the yaml base config which makes the conf much cleaner. Thanks for the suggested. I also added the licence info from MS.

**Comment by okuchaiev** [comment]
@yidong72 we really need docs/tutorial on this. Do you want to send them as separate PR so that this one can make code-freeze?

**Comment by yidong72** [comment]
@okuchaiev , yes. let's merge this first. I can work on the tutorial for the next sprint.

**Review by MaximumEntropy — APPROVED** [review]
LGTM!
